src/request.rs

```diff
@@ -14,28 +14,113 @@ use super::SerdeValue;
             write!(f, "{}", url.as_str())
         } else {
-            write!(f, "{:?}", self.0)
+            write!(f, "{:?}", self.result)
         }
     }
 }
 
 impl From<String> for ParsedUrl {
     fn from(s: String) -> Self {
-        ParsedUrl(s.parse())
+        let result = s.parse();
+        let origin = if result.is_err() { Some(s) } else { None };
+        ParsedUrl::new(result, origin)
     }
 }
 
 impl From<Url> for ParsedUrl {
     fn from(url: Url) -> Self {
-        ParsedUrl(Ok(url))
+        ParsedUrl::new(Ok(url), None)
+    }
+}
+
+impl ParsedUrl {
+    fn new(mut result: std::result::Result<Url, ParseError>, origin: Option<String>) -> Self {
+        if let Ok(url) = &result {
+            if url.host_str().is_none() {
+                // No hostname is fine for urls in general, but not for website urls.
+                result = Err(ParseError::EmptyHost);
+            }
+        }
+
+        ParsedUrl { result, origin }
+    }
+
+    // NOTICE:
+    // The unwrap() in all folllowing functions is ok because
+    // we can only get a `ParsedUrl` via `Request::parsed_url()`, in
+    // which case a potential error state is surfaced. In other words,
+    // if the user has an instance of ParsedUrl, they are guaranteed it
+    // contains an Ok() value.
+
+    /// Handle the parsed url as a standard [`url::Url`].
+    pub fn as_url(&self) -> &Url {
+        self.result.as_ref().unwrap()
+    }
+
+    /// Get the scheme of the parsed url, i.e. "https" or "http".
+    pub fn scheme(&self) -> &str {
+        self.as_url().scheme()
+    }
+
+    /// Host of the parsed url.
+    pub fn host(&self) -> &str {
+        // this unwrap() is ok, because ParsedUrl is tested for empty host
+        // urls in ParsedUrl::new(). At this point we are guaranteed there
+        // is no empty host.
+        self.as_url().host_str().unwrap()
+    }
+
+    /// Port of the parsed url, if available. Ports are only available if they
+    /// are present in the original url. Specifically the scheme default ports,
+    /// 443 for `https` and and 80 for `http` are `None` unless explicitly
+    /// set in the url, i.e. `https://my-host.com:443/some/path`.
+    pub fn port(&self) -> Option<u16> {
+        self.as_url().port()
+    }
+
+    /// Path of the parsed url.
+    pub fn path(&self) -> &str {
+        self.as_url().path()
+    }
+
+    /// Returns all query parameters as a vector of key-value pairs.
+    ///
+    /// ```
+    /// # fn main() -> Result<(), ureq::Error> {
+    /// # ureq::is_test(true);
+    /// let req = ureq::get("http://httpbin.org/get")
+    ///     .query("foo", "42")
+    ///     .query("foo", "43");
+    ///
+    /// assert_eq!(req.parsed_url().unwrap().query_pairs(), vec![
+    ///     ("foo".to_string(), "42".to_string()),
+    ///     ("foo".to_string(), "43".to_string())
+    /// ]);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn query_pairs(&self) -> Vec<(String, String)> {
```

Comment:
I propose to make this return `Vec<(&str, &str)>`, because the caller doesn't generally need owned strings. `&str` is more consistent with the rest of our API, and leaves the door open to a more optimized implementation later. In #348 I did this by storing a copy of the query pairs in the Url object on construction.

Reply:
Ideally I want `Vec<(&str, &str)>`, but, as you point out, that forces us to do the String allocation on construction, effectively penalizing every use of `Request`, also those that don't want this.

However, an escape hatch could be to only do this when someone calls `parsed_url()`. I'll look into it.

Reply:
Done